src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/CaptureGroups.java

```diff
@@ -0,0 +1,129 @@
+package io.stargate.sgv2.jsonapi.service.resolver.matcher;
+
+import io.stargate.sgv2.jsonapi.api.model.command.Command;
+import io.stargate.sgv2.jsonapi.service.operation.query.DBFilterLogicalExpression;
+import java.util.*;
+import java.util.function.BiConsumer;
+
+/**
+ * All the {@link CaptureGroup}s we got from matching against a command.
+ *
+ * <p>This is "result" of running the FilterMatcher, and the value we pass to the resolver function
+ * so it has raw command and all the groups.
+ *
+ * <p>Each Capture you create from {@link FilterMatcher#capture(Object)} with a Marker is available
+ * here as a {@link CaptureGroup} via {@link #getGroup(Object)}.
+ *
+ * <p>Created in the {@link FilterMatcher} via a builder
+ *
+ * <p>T - The {@link Command} that is filtered against
+ *
+ * <p>Since we need to keep the logical relation information when doing captures for
+ * LogicalExpression, so CaptureGroups needs to have a recursive defined structure
+ */
+public class CaptureGroups<T extends Command> {
+
+  private DBFilterLogicalExpression.DBLogicalOperator dbLogicalOperator;
+  private final T command;
+
+  public final Map<Object, CaptureGroup<?>> captureGroupMap;
+
+  private final List<CaptureGroups<T>> captureGroupsList;
+
+  public CaptureGroups(T command, DBFilterLogicalExpression.DBLogicalOperator dbLogicalOperator) {
+    this.command = command;
+    this.dbLogicalOperator = dbLogicalOperator;
+    this.captureGroupMap = new HashMap<>();
+    this.captureGroupsList = new ArrayList<>();
+  }
+
+  /**
+   * Get the {@link CaptureGroup} that has the result of the {@link FilterMatcher.Capture} created
+   * with the supplied marker, construct one if absent
+   *
+   * @param marker
+   * @return CaptureGroup
+   */
+  public CaptureGroup<?> getGroup(Object marker) {
+    return captureGroupMap.computeIfAbsent(marker, f -> new CaptureGroup(new HashMap<>()));
+  }
+
+  /**
+   * Get the {@link CaptureGroup} that has the result of the {@link FilterMatcher.Capture} created
+   * with the supplied marker. If there is no target marker in captureGroup map, return an empty
+   * optional
+   *
+   * @param marker
+   * @return Optional<CaptureGroup>
+   */
+  public Optional<CaptureGroup<?>> getGroupIfPresent(Object marker) {
```

Comment:
This would be easier to use if it was:

```java
consumeGroupIfPresent(Object marker, Consumer<CaptureExpression<DataType>>>)
```

would simplify the repetative pattern below 
```java
          captureGroups
              .getGroupIfPresent(DYNAMIC_DOCID_GROUP)
              .ifPresent(
                  captureGroup -> {
                    CaptureGroup<Object> dynamicDocIDGroup = (CaptureGroup<Object>) captureGroup;
                    dynamicDocIDGroup.consumeAllCaptures(
```

Reply:
this is not a quick fix, cause lots of noise in the class.
Let me create a ticket, and do a followup. This has no functional impact.